test/test_foreach.py

```diff
@@ -698,57 +738,44 @@ def test_bool_scalarlist(self, device, dtype):
+                            foreach_bin_op(tensors, scalars)
                     else:
                         if self.device_type == "cpu":
                             expected = [torch_bin_op(t, s) for t, s in zip(tensors, scalars)]
                             res = foreach_bin_op(tensors, scalars)
 
-                            # see TODO[Fix scalar list]
-                            if dtype in torch.testing.integral_types():
-                                self.assertEqual(res, [e.to(torch.float32) for e in expected])
-                            else:
-                                self.assertEqual(res, expected)
+                            self.assertEqual(res, expected)
 
-                            if dtype in torch.testing.integral_types():
+                            if dtype in torch.testing.integral_types() and foreach_bin_op_ == torch._foreach_div_:
                                 with self.assertRaisesRegex(RuntimeError, "result type Float can't be cast to the desired "):
                                     foreach_bin_op_(tensors, scalars)
                             else:
                                 foreach_bin_op_(tensors, scalars)
                                 self.assertEqual(tensors, expected)
                         else:
-                            expected = [torch_bin_op(t, s) for t, s in zip(tensors, scalars)]
-                            res = foreach_bin_op(tensors, scalars)
+                            if foreach_bin_op == torch._foreach_sub:
+                                with self.assertRaisesRegex(RuntimeError, "Subtraction, the `-` operator, with a bool tensor is not supported."):
+                                   expected = [torch_bin_op(t, s) for t, s in zip(tensors, scalars)]
 
-                            if dtype in torch.testing.integral_types():
-                                self.assertEqual(res, [e.to(dtype) for e in expected])
+                                res = foreach_bin_op(tensors, scalars)
+                                foreach_bin_op_(tensors, scalars)
+                                self.assertEqual(res, tensors)
                             else:
-                                self.assertEqual(res, expected)
+                                expected = [torch_bin_op(t, s) for t, s in zip(tensors, scalars)]
+                                res = foreach_bin_op(tensors, scalars)
 
-                            foreach_bin_op_(tensors, scalars)
-                            self.assertEqual(res, tensors)
+                                if dtype in torch.testing.integral_types():
+                                    self.assertEqual(res, [e.to(dtype) for e in expected])
+                                else:
+                                    self.assertEqual(res, expected)
+
+                                foreach_bin_op_(tensors, scalars)
+                                self.assertEqual(res, tensors)
```

Comment:
Some general feedback is: the tests are really hard to read because they branch so much. I'm not sure how to make them easier to read, though. Perhaps we should consider splitting them up or think about why the behavior diverges so much?

Reply:
Yes, i fully agree. thats why in the next PR in this stack i separate all the tests. Much easier to read and maintain.